core/src/main/scala/org/apache/spark/deploy/history/HistoryServerMemoryManager.scala

```diff
@@ -0,0 +1,82 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.deploy.history
+
+import java.util.concurrent.atomic.AtomicLong
+
+import scala.collection.mutable.HashMap
+
+import org.apache.spark.SparkConf
+import org.apache.spark.internal.Logging
+import org.apache.spark.internal.config.History._
+import org.apache.spark.util.Utils
+
+/**
+ * A class used to keep track of in-memory store usage by the SHS.
+ */
+private class HistoryServerMemoryManager(
+    conf: SparkConf) extends Logging {
+
+  private val maxUsage = conf.get(MAX_IN_MEMORY_STORE_USAGE)
+  private val currentUsage = new AtomicLong(0L)
+  private val active = new HashMap[(String, Option[String]), Long]()
+
+  def initialize(): Unit = {
+    logInfo("Initialized memory manager: " +
+      s"current usage = ${Utils.bytesToString(currentUsage.get())}, " +
+      s"max usage = ${Utils.bytesToString(maxUsage)}")
+  }
+
+  def lease(
+      appId: String,
+      attemptId: Option[String],
+      eventLogSize: Long,
+      isCompressed: Boolean): Unit = {
+    val memoryUsage = approximateMemoryUsage(eventLogSize, isCompressed)
+    if (memoryUsage + currentUsage.get > maxUsage) {
+      throw new RuntimeException("Not enough memory to create hybrid store " +
+        s"for app $appId / $attemptId.")
+    }
+    active.synchronized {
+      active(appId -> attemptId) = memoryUsage
+    }
+    currentUsage.addAndGet(memoryUsage)
+    logInfo(s"Leasing ${Utils.bytesToString(memoryUsage)} memory usage for " +
+      s"app $appId / $attemptId")
+  }
+
+  def release(appId: String, attemptId: Option[String]): Unit = {
+    val memoryUsage = active.synchronized { active.remove(appId -> attemptId) }
+
+    memoryUsage match {
+      case Some(m) =>
+        currentUsage.addAndGet(-m)
+        logInfo(s"Released ${Utils.bytesToString(m)} memory usage for " +
+          s"app $appId / $attemptId")
+      case None =>
+    }
+  }
+
+  private def approximateMemoryUsage(eventLogSize: Long, isCompressed: Boolean): Long = {
+    if (isCompressed) {
+      eventLogSize * 2
```

Comment:
how accurate is this in practice? Will depend on the codec chosen. did you try some large files?

Reply:
This memory usage approximation follows the way we estimate disk usage in HistoryServerDiskManager. In disk manager, we estimate the disk usage based on the log file size. I did some experiments to compare the memory usage of hybridStore (measured through org.apache.spark.util.SizeEstimator) and the disk usage of levelDB on a set of log files. I found they are similar. So I think if disk usage approximation works, this kind of memory usage approximation should work.

Reply:
the difference is though that is disk usage, not memory usage. If you go over disk usage temporarily generally not as big of a deal if you go over memory usage.  I'm wondering for instance if the codec is Zstd could the memory usage be much more if it compressed it very well.  I'm not sure a good solution to this at the moment other then perhaps make the multiplication factor configurable, but I'm fine with waiting and seeing how it runs and if you have tested some cases and seems close that is fine to start with.

Please add the comments saying you copied from the HistoryServerDiskManager and put similar comments that function has about it being approximate.

Reply:
I will add comments about this function and test compressed files later. So far I think this estimation is conservative and safe. For large uncompressed log files, the memory and disk usage are usually much smaller than the estimated usage(half of the log file size). I measured some log files with size ~1g that I randomly picked from production-use spark jobs.

log filesize | 1.0g | 1.26g | 1.16g | 1.11g
-- | -- | -- | -- | --
memory usage | 113m | 149m | 63m | 46m
leveldb filesize | 175m | 211m | 42m | 77m

Reply:
comparison between memory usage and disk usage for some randomly chosen 2.4 logs, parsed by 2.4 SHS

log size | 1.0g | 1.26g | 1.16g | 1.11g
-- | -- | -- | -- | --
memory usage | 97.8m | 232m | 53.1m | 31.4m
leveldb filesize | 183M | 368m | 220m | 99m

Reply:
It'd be nice if we experiment with all available compressions for event log and have numbers here.

Reply:
Sure I will do the experiment with different compressions and put the number here.

Reply:
any update on the numbers?

Reply:
Sorry for the delay, I will update it soon.

Reply:
Hi, I updated the number on https://github.com/apache/spark/pull/28412#issuecomment-653110422.